src/RFC-0321_ProcessingForeignProposals.md

```diff
@@ -0,0 +1,113 @@
+
+## Language
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",
+"NOT RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in
+[BCP 14](https://tools.ietf.org/html/bcp14) (covering RFC2119 and RFC8174) when, and only when, they appear in all capitals, as
+shown here.
+
+## Disclaimer
+
+This document and its content are intended for information purposes only and may be subject to change or update
+without notice.
+
+This document may include preliminary concepts that may or may not be in the process of being developed by the Tari
+community. The release of this document is intended solely for review and discussion by the community of the
+technological merits of the potential system outlined herein.
+
+## Goals
+
+This RFC describes the process of distributing and processing foreign proposals in the Tari DAN Cerberus Model
+
+Across the entire network transactions must be processed or time out. When a transaction is started on a shard, it locks up substates, preventing other transactions from completing. Therefore if a transaction is started on a shard, it 
+should complete or be aborted in a timely manner to release the resources.
+
+Ordering of transactions is also important across all shards. While the different
+shards form a Direct Acyclic Graph (DAG), there should be a consistent ordering 
+across the whole network.
+
+
+
+## Related Requests for Comment
+
+<!-- * [RFC-0111: Base Node Architecture](./RFC-0111_BaseNodeArchitecture.md) -->
+
+## Glossary
+
+* Block - A second layer block, consisting of ordered commands
+* Command - Command can either be Prepare, LocalPrepared, Accept, and moves a transaction into that state.
+
+## Description
+
+To solve the above problems, we'll use reliable broadcast and process foreign evidence in order.
+
+In a local shard committee, the proposed block **must** include a counter per epoch and per shard. If the proposal includes transactions that involve other shards, this counter **must** be incremented.
+
+When the proposed block becomes committed locally, the block **must** be broadcast to each involved shard that was incremented, along with evidence of being committeed (The chain of QCs must be included).
+
+To ensure this, each node in the local committee will forward this committed block, along with a 3 chain of QC's proving it was committeed.
+
+As a local committee member, when I receive a foreign proposal, if it is valid I will queue up a special command ForeignProposal(number, QC_Hash) that I must propose 
+when I am next leader (if it has not been proposed already). I also **should** request all transaction hashes that I have not seen from involved_shards for each transaction in the proposal, and add them to my mempool for execution.
+
+In the same order of transactions in the proposal, the transactions must be sequenced into the local chain as either a TIMEOUT or a LOCALPREPARE(TxId, ForeignShardId).
```

Comment:
I think the ordering of transactions within a proposal will cause some inconsistency.
E.g. I have Shard Sa and Sb, and Sa receive Ta (transaction over both shards) and Sb receive Tb (also over both shards). Then Sa will prepare a block with ordering Ta,Tb and Sb will prepare a block with order Tb,Ta. Then the foreign proposal will have wrong order compared to local proposal.

Why does the order in the proposal matters? It's not like I can have concurrent transactions in the proposal, right?

Reply:
I suppose the order inside the proposal does not matter, so long as the proposals themselves are processed in entirety before any transactions for a new proposal can be sequenced. The rule should then be updated that any transactions that have not been processed within the time limit must be timed out

Reply:
I'm curious when can it happen that I receive a foreign proposal with some Tx and I'm not able to do localprepare(tx). That means that even the committee from which I got the foreign proposal don't have the Tx.
Now I understand why you want timeout in foreign committee to send you that it timed out.